### ISSUE TITLE
Line output does not match traceback output

### ISSUE THREAD

**Issue by spacemanspiff2007** [alexmojaki/stack_data#38]
I've observed a very strange phenomenon today.
Additionally to the prettified traceback I include the normal traceback since the stack data is only generated for some files.
The line numbers were both times the same, but the shown code (the line content) of stack_data was wrong.
I've included an additional `assert` statement and the python code gets reloaded during runtime with `runpy.run_path`

What stack_data produces for line 68:
```
--------------------------------------------------------------------------------
      64 |     i = self.out_rgb
      65 |     r, g, b = self.out_rgb.get_rgb(max_rgb_value=100)
      66 |     self.logger.debug(f'Color: {i.hue:3.0f},{i.saturation:3.0f},{i.brightness:3.0f} ({r:3.0f},{g:3.0f},{b:3.0f})')
 -->  68 | async def leistung_change(self, event: ValueChangeEvent):
      69 |     self.logger.debug(f'{self.power.value:4.1f}W')
    ------------------------------------------------------------
```

Traceback for line 68:
```
 Traceback (most recent call last):
   File "/opt/habapp/rules/code.py", line 68, in rgb_out_change
     assert isinstance(value, tuple) or value is 'ON' or value is 'OFF'
 AssertionError
```

Code that produced the exception (line 68 is the assert statement)
```python
    async def rgb_out_change(self, event: ValueChangeEvent):
        value = event.value
        assert isinstance(value, tuple) or value == 'ON' or value == 'OFF'
        self.out_rgb.oh_send_command(value)

    async def dimmer_rgb_change(self, event: ValueChangeEvent):
        i = self.out_rgb
        r, g, b = i.get_rgb()
        self.logger.debug(f'Color: {i.hue:3.0f},{i.saturation:3.0f},{i.brightness:3.0f} ({r:3.0f},{g:3.0f},{b:3.0f})')
```

Is there some kine of line cache that stores the line content?

**Comment by alexmojaki** [comment]
> Is there some kine of line cache that stores the line content?

Yes there is, this is the result of that cache becoming stale after reloading the file.

Try upgrading `executing`, I recently released a new version that refreshes the cache after `importlib.reload`. But I don't know if that will help here.

**Comment by spacemanspiff2007** [comment]
Hmm - `linecache` seems to get properly invalidated.
Is there a possibility to add
- the file time and file size from `linecache.cache` to your cache and check it before formatting the exception. That way both caches would be kept in sync
- a function I can call that invalidates the cache

But first I'll try to generate an example to reproduce the issue to see if it gets fixed by the upgraded `executing`.
Maybe it's already fixed.

**Comment by spacemanspiff2007** [comment]
I tried to use the original format code so this might a bit too complex.
This example creates a file next to the source code, executes it.
Then the file gets overwritten and executed again.

```python
import ast
from inspect import isclass
from pathlib import Path
from runpy import run_path
from traceback import format_exception as _format_exception
from typing import List, Tuple, Callable, Set, Union, Any

from stack_data import FrameInfo, LINE_GAP, Options, Variable

# -----------------------------------------------------------------------------------------------------------
# Exception formatting code
# -----------------------------------------------------------------------------------------------------------

PRE_INDENT = 4
SEPARATOR_NEW_FRAME = '-' * 80
SEPARATOR_VARIABLES = ' ' * (PRE_INDENT - 1) + '-' * 60


ORDER_VARIABLE: Tuple[Callable[[Variable], bool], ...] = (
    lambda x: isclass(x.value),
)


def format_frame_variables(tb: List[str], stack_variables: List[Variable]):
    if not stack_variables:
        return None

    # remove variables that shall not be printed
    used_vars: List[Variable] = [v for v in stack_variables]

    # attributes
    dotted_names: Set[str] = {n.name.split('.')[0] for n in used_vars if '.' in n.name}

    # Sort output
    used_vars = sorted(used_vars, key=lambda x: (
        isinstance(x.nodes[0], ast.Compare),                                        # Compare objects last
        not any(map(
            lambda y: x.name == y or x.name.startswith(y + '.'), dotted_names)),    # Classes with attributes
        x.name.lower()                                                              # Name lowercase
    ))

    variables = {}

    # variables by order
    for add_var in ORDER_VARIABLE:
        for var in used_vars:
            name = var.name
            if name in variables:
                continue
            if add_var(var):
                variables[name] = var.value

    # rest of the variables
    for var in used_vars:
        if var.name not in variables:
            variables[var.name] = var.value

    if not variables:
        return None

    # Add variables to traceback
    tb.append(SEPARATOR_VARIABLES)

    for name, value in variables.items():
        tb.append(f'{" " * (PRE_INDENT + 1)}{name} = {repr(value)}')

    tb.append(SEPARATOR_VARIABLES)


def format_frame_info(tb: List[str], frame_info: FrameInfo, is_last=False) -> bool:
    filename = frame_info.filename

    # calc max line nr for indentation
    max_line = frame_info.lineno + frame_info.options.after
    # it's possible that his list is empty (e.g. if we execode in sphinx-exec-code)
    if frame_info.lines:
        max_line = frame_info.lines[-1].lineno

    indent = len(str(max_line)) + 1

    tb.append(f'File "{filename}", line {frame_info.lineno} in {frame_info.code.co_name}')
    tb.append(SEPARATOR_NEW_FRAME)
    for line in frame_info.lines:
        if line is LINE_GAP:
            tb.append(f"{' ' * (PRE_INDENT + indent - 1)}(...)")
        else:
            tb.append(f"{'-->' if line.is_current else '':{PRE_INDENT}s}{line.lineno:{indent}d} | {line.render()}")

    format_frame_variables(tb, frame_info.variables)
    tb.append('')
    return True


def append_short_traceback(tb: List[str], e: Union[Exception, Tuple[Any, Any, Any]]):
    for line in _format_exception(*e) if isinstance(e, tuple) else _format_exception(type(e), e, e.__traceback__):
        for sub_lines in line.splitlines():
            tb.append(sub_lines.rstrip())


DEFAULT_OPTIONS = Options(include_signature=True, max_lines_per_piece=5)


def format_exception(e: Union[Exception, Tuple[Any, Any, Any]]) -> List[str]:
    tb: List[str] = []

    all_frames = tuple(FrameInfo.stack_data(e[2] if isinstance(e, tuple) else e.__traceback__, DEFAULT_OPTIONS))
    last_frame = len(all_frames) - 1

    added = True
    for i, frame_info in enumerate(all_frames):
        if isinstance(frame_info, FrameInfo):
            added = format_frame_info(tb, frame_info, is_last=i == last_frame)
        else:
            # repeated frames in case of recursion
            if added:
                tb.append(f"... {frame_info.description} ...\n")

    # add a short traceback
    tb.append(SEPARATOR_NEW_FRAME)
    append_short_traceback(tb, e)

    return tb


# -----------------------------------------------------------------------------------------------------------
# Code execution code
# -----------------------------------------------------------------------------------------------------------

def run_code(code: str, name: str):
    tmp_file = Path(__file__).with_name('__test_run_py.py')
    tmp_file.write_text(code)

    glob = run_path(str(tmp_file))
    glob = {k: v for k, v in glob.items() if not k.startswith('_')}  # remove entries so the traceback isn't

    try:
        glob[name]()
    except Exception as e:
        print("\n".join(format_exception(e)))

    print('-' * 120)
    print('-' * 120)


code_1 = """
def func():
    print('Line 1')
    print('Line 2')
    a = 1
    1/0

"""


run_code(code_1, 'func')

code_2 = """
def func1():
    print('Line 1')
    print('Line 2')
    print('Line 3')
    print('Line 4')
    print('Line 5')
    a = 1
    1/0

"""

run_code(code_2, 'func1')
```

The first call to `run_code` does format the exception as expected.

```
--------------------------------------------------------------------------------
     2 | def func():
     3 |     print('Line 1')
     4 |     print('Line 2')
     5 |     a = 1
-->  6 |     1/0
   ------------------------------------------------------------
     a = 1
   ------------------------------------------------------------
```

I would have expected the second call to hit the cache but instead I'm seeing another exception.

```
Traceback (most recent call last):
  File "\src\__test_trace_format.py", line 161, in <module>
    run_code(code_2, 'func1')
  File "\src\__test_trace_format.py", line 131, in run_code
    print("\n".join(format_exception(e)))
  File "\src\__test_trace_format.py", line 108, in format_exception
    added = format_frame_info(tb, frame_info, is_last=i == last_frame)
  File "\src\__test_trace_format.py", line 72, in format_frame_info
    if frame_info.lines:
  File "\venv310\lib\site-packages\stack_data\utils.py", line 145, in cached_property_wrapper
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "\venv310\lib\site-packages\stack_data\core.py", line 698, in lines
    pieces = self.included_pieces
  File "\venv310\lib\site-packages\stack_data\utils.py", line 145, in cached_property_wrapper
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "\venv310\lib\site-packages\stack_data\core.py", line 649, in included_pieces
    pos = scope_pieces.index(self.executing_piece)
  File "\venv310\lib\site-packages\stack_data\utils.py", line 145, in cached_property_wrapper
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "\venv310\lib\site-packages\stack_data\core.py", line 628, in executing_piece
    return only(
  File "\venv310\lib\site-packages\executing\executing.py", line 164, in only
    raise NotOneValueFound('Expected one value, found 0')
executing.executing.NotOneValueFound: Expected one value, found 0

Process finished with exit code 1
```

I'm unsure how to provide a fix for it and what causes the exception.

**Comment by alexmojaki** [comment]
Released `executing` 1.0.0 that fixes this.

**Comment by spacemanspiff2007** [comment]
I can confirm that the issue fixed
